File: mobile/ios/OpenU60/Features/RouterSettings/VPN/VPNPassthroughViewModel.swift
import SwiftUI

@Observable
@MainActor
final class VPNPassthroughViewModel {
    var config: VPNPassthroughConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    var editL2tp: Bool = false
    var editPptp: Bool = false
    var editIpsec: Bool = false

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let data = try await client.getJSON("/api/router/vpn")
            config = VPNPassthroughParser.parse(data)
            editL2tp = config.l2tp
            editPptp = config.pptp
            editIpsec = config.ipsec
        } catch {
            showMessage("Failed to load VPN: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func apply() async {
        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/vpn", body: [
                "l2tp_passthrough": editL2tp ? "1" : "0",
                "pptp_passthrough": editPptp ? "1" : "0",
                "ipsec_passthrough": editIpsec ? "1" : "0"
            ])
            showMessage("VPN passthrough updated", isError: false)
            config = VPNPassthroughConfig(l2tp: editL2tp, pptp: editPptp, ipsec: editIpsec)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/vpn/VPNPassthroughViewModel.kt
package com.openu60.feature.router.vpn

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.VPNPassthroughConfig
import com.openu60.core.model.VPNPassthroughParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class VPNPassthroughState(
    val config: VPNPassthroughConfig = VPNPassthroughConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class VPNPassthroughViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(VPNPassthroughState())
    val state: StateFlow<VPNPassthroughState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/vpn/passthrough")
                val config = VPNPassthroughParser.parse(data)
                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun toggle(field: String, value: Boolean) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/vpn/passthrough", mapOf(field to if (value) "1" else "0"))
                _state.value = _state.value.copy(message = "Updated", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) toggle(field, value) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}
